### PR TITLE
Make `unused_allocation` lint against `Box::new` too

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Previously it only linted against `box` syntax, which likely won't ever be stabilized, which is pretty useless. Even now I'm not sure if it's a meaningful lint, but it's at least something :shrug:

This means that code like the following will be linted against:
```rust
Box::new([1, 2, 3]).len();
f(&Box::new(1)); // where f : &i32 -> ()
```
The lint works by checking if a `Box::new` (or `box`) expression has an a borrow adjustment, meaning that the code that first stores the box in a variable won't be linted against:
```rust
let boxed = Box::new([1, 2, 3]); // no lint
boxed.len();
```

### CODE REVIEW

**Comment by nagisa** [comment]
This unfortunately may need a full FCP at least. While the lint itself is not new, the way it is expanded in scope is going to be going from affecting approximately no users to affecting a large majority of them. If there is `-Dwarnings` anywhere it would also prevent the codebase from compiling.

I guess one thing we can do before an FCP is run a crater. Could you make this lint a error-by-default so we can kick-off a run and see how much of the ecosystem actually explodes?

**Comment by WaffleLapkin** [comment]
The FCP makes sense, that's a significant change to the lint. However I don't really see the reasoning for running the crater. Using `#[deny(warning)]` is an anti-pattern and we never promised that the lints are stable...

**Comment by Nilstrieb** [comment]
@nagisa making an FCP sounds reasonable, can you start it? I don't think running crater makes sense as this won't affect dependencies (cap-lints) and projects _want_ breakage on new warnings if they deny warnings so it's fine.

**Comment by apiraino** [comment]
I'll label the issue to signal that an FCP is needed (hope I'm using the correct incantation).

@rustbot label -S-waiting-on-review S-waiting-on-team needs-fcp

**Comment by oli-obk** [comment]
The changes of this PR are documented in the main post. TLDR: we're making a lint smarter, which means it will trigger on code it previously did not trigger on.

@rfcbot merge

**Comment by est31** [comment]
I have done the same thing that this PR is doing in #97774 , but eventually closed it, due to the lint suggesting `assert!(a == &8)` (which doesn't work) for  `assert!(a == Box::new(8))`.

But in retrospect, this seems like a papercut to me: the lint does not emit any rustfixable suggestions or anything. I think this PR is good because it makes the lint less specific to box syntax, so :+1: from me.

**Comment by Nilstrieb** [comment]
ping @nikomatsakis @pnkfelix @wesleywiser :)

File: library/alloc/src/tests.rs [hunk 1]
```diff
@@ -15,32 +14,25 @@ fn test_owned_clone() {
...
+    let b: Box<Test> = b.downcast::<Test>().unwrap();
+    assert_eq!(*b, Test);
 
     let a = Box::new(8) as Box<dyn Any>;
     let b = Box::new(Test) as Box<dyn Any>;
 
-    assert!(a.downcast::<Box<Test>>().is_err());
-    assert!(b.downcast::<Box<i32>>().is_err());
+    assert!(a.downcast::<Box<i32>>().is_err());
+    assert!(b.downcast::<Box<Test>>().is_err());
```
Comment (pnkfelix) [inline library/alloc/src/tests.rs:35]: It took me a fair amount of squinting before I realized why this change was an improvement to the test. :)

**Comment by Nilstrieb** [comment]
Now that you also lint on `Box::new`, can you use that in the example in `compiler/rustc_lint/src/unused.rs L1352`?

**Comment by WaffleLapkin** [comment]
@Nilstrieb I'm not sure what you mean... I've [already done this](https://github.com/rust-lang/rust/pull/104363/files#diff-e25b2c6c1fa434ab8078db81003888370f06b9da5503d1590757350be8e31103R1353)...

File: tests/ui/lint/unused/unused-allocation.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#![feature(box_syntax)]
+#![deny(unused_allocation)]
+
+fn main() {
+    _ = (box [1]).len(); //~ error: unnecessary allocation, use `&` instead
```
Comment (Nilstrieb) [inline tests/ui/lint/unused/unused-allocation.rs:5]:
```suggestion
#![feature(rustc_attrs)]
#![deny(unused_allocation)]

fn main() {
    _ = (#[rustc_box] Box::new([1])).len(); //~ error: unnecessary allocation, use `&` instead
```

**Review by Noratrieb — COMMENTED** [review]
r=me after the FCP is complete and the test is changed
